### PR TITLE
Added infinite nesting to TOC

### PR DESCRIPTION
**Pull request by kappu72**

### CODE REVIEW

File: web/client/utils/LayersUtils.js [hunk 1]
```diff
@@ -33,58 +53,53 @@ var LayersUtils = {
...
     reorder: (groups, allLayers) => {
         return allLayers.filter((layer) => layer.group === 'background')
-            .concat(initialReorderLayers(groups, allLayers));
+            .concat(reorderLayers(groups, allLayers));
     },
     denormalizeGroups: (allLayers, groups) => {
-
         let getGroupVisibility = (nodes) => {
-            let visibility = true;
+            let visibility = false;
```
Comment (mbarto) [inline web/client/utils/LayersUtils.js]: I don't think this refactor is correct. The group visibility is true only if every node is visible

File: web/client/utils/LayersUtils.js [hunk 2]
```diff
@@ -9,21 +9,41 @@
 const assign = require('object-assign');
 const {isObject, isArray} = require('lodash');
 
+const getGroup = (groupId, groups) => {
+    return groups.filter((subGroup) => isObject(subGroup) && subGroup.id === groupId).pop();
```
Comment (mbarto) [inline web/client/utils/LayersUtils.js]: use lodash head to get first element of array, not pop

File: web/client/reducers/layers.js [hunk 3]
```diff
@@ -204,7 +149,29 @@ function layers(state = [], action) {
             });
             let originalNode = head(flatLayers.filter((layer) => { return (layer[selector] === action.node || layer[selector].indexOf(action.node + '.') === 0); }));
             if (!sameGroup && originalNode ) {
-                let newGroups = moveNode(state.groups, action.node, (action.options.group || 'Default'));
+                // Remove layers from old group
```
Comment (mbarto) [inline web/client/reducers/layers.js]: Why did you remove moveNode?

File: web/client/utils/LayersUtils.js (hunk 1 above)
Comment (kappu72) [inline web/client/utils/LayersUtils.js]: Fixed

File: web/client/utils/LayersUtils.js (hunk 2 above)
Comment (kappu72) [inline web/client/utils/LayersUtils.js]: fixed

File: web/client/reducers/layers.js (hunk 3 above)
Comment (kappu72) [inline web/client/reducers/layers.js]: Fixed

File: web/client/utils/LayersUtils.js (hunk 2 above)
Comment (kappu72) [inline web/client/utils/LayersUtils.js]: The id is autogenerated and is the same then before, are name and title that change.

File: web/client/components/TOC/DefaultLayerOrGroup.jsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright 2016, GeoSolutions Sas.
```
Comment (kappu72) [inline web/client/components/TOC/DefaultLayerOrGroup.jsx]: fixed

File: web/client/components/TOC/fragments/GroupTitle.jsx [hunk 5]
```diff
@@ -31,7 +31,7 @@ const GroupTitle = React.createClass({
         let expanded = (this.props.node.expanded !== undefined) ? this.props.node.expanded : true;
         let groupTitle = this.props.node && this.props.node.title || 'Default';
         return (
-            <div className="toc-group-title" onClick={() => this.props.onClick(this.props.node.name, expanded)} style={this.props.style}>
+            <div className="toc-group-title" onClick={() => this.props.onClick(this.props.node.id, expanded)} style={this.props.style}>
```
Comment (kappu72) [inline web/client/components/TOC/fragments/GroupTitle.jsx]: Yes,  group id is mandatory, but is autogenerated in LayersUtils.getLayersByGroup.

File: web/client/reducers/layers.js [hunk 6]
```diff
@@ -32,79 +34,23 @@ const deepChange = (nodes, findValue, propName, propValue) => {
...
-};
-
 const getNode = (nodes, name) => {
     if (nodes && isArray(nodes)) {
         return nodes.reduce((previous, node) => {
             if (previous) {
                 return previous;
             }
-            if (node && node.name === name) {
+            if (node && (node.name === name || node.id === name)) {
```
Comment (kappu72) [inline web/client/reducers/layers.js]: fixed

File: web/client/reducers/layers.js [hunk 7]
```diff
@@ -116,6 +62,32 @@ const getNode = (nodes, name) => {
...
+    // Remove node from old group
+    let newGroups = deepRemove(groups, node);
+    // Check if group to move to exists
+    let group = getNode(newGroups, groupId);
+    if (!group) {
+        // Create missing group
+        group = head(LayersUtils.getLayersByGroup([getNode(newLayers, node)]));
+        // check for parent group if exist
+        const parentGroup = groupId.split('.').reduce((tree, gName, idx) => {
+            const gId = groupId.split(".", idx + 1).join('.');
```
Comment (kappu72) [inline web/client/reducers/layers.js]: parentGroup is the object parentGroup.id is the parentId

**Comment by offtherailz** [comment]
Should we use `/` instead of `.` as a separator for groups? Does it make sense?
As a plus we should allow to escape the group separator somehow
